posts/2017-06-02-devops-serverless-variables.md

```diff
@@ -0,0 +1,155 @@
+
+Let’s take a look at an example. Say you want to control which AWS account to use based on certain logical conditions. You can make use of the variable system to setup the correct value for AWS_PROFILE environment variable for you before deployment. All you have to do is just reference a JS variables file that contain that logic. So in serverless.yml
+
+
+```yml
+service: myservice
+custom:
+  setup: ${file(vars.js):setup} # this is just to trigger the code. This var is not needed at all!
+```
+Then in vars.js, you can write your logic:
+
+
+```
+module.exports.setup = () => {
+  /*
+   * some logic here...
+   */
+  process.env.AWS_Profile = “sam”;
+  return process.env.AWS_Profile; // you don’t even need to return because you’d never use that value!
+};
+```
+This way, vars.js could determine which AWS_Profile to use.
+
+## Organizing Your Serverless Configuration
+If you’ve been using Serverless in production for a while now, most likely you’ve realized how complex your serverless.yml can become. Many of the configuration options could actually be reused, not only within your services, but throughout all your services.
+
+As the framework matured, we started to notice patterns in how users are organizing their config file with the help of the variable system. Here are some common structures:
+
+### Custom Resources Files
+Real world serverless applications require the use of many AWS resources that you’ll need to set up in the `resources` section of `serverless.yml`. Those resources themselves have many configurations on their own and they will most likely reference other variables. To keep `serverless.yml` from getting bloated, many users keep all their AWS resources in a single `resources.json` file, and reference that file in `serverless.yml` like this:
+
+```yml
+Resources:
+  resources: ${file(resources.json)}
+```
+You can also have an outputs.json file and reference it the same way:
+
+```yml
+Resources:
+  resources: ${file(resources.json)}
+  outputs: ${file(outputs.json)}
+```
+Then, inside of your resources.json & outputs.json, you can still reference other variables as required by your resources.
+
+### Function Config Files
+Use cases exist where users make the decision to follow a monolithic architecture when working the Serverless. That means they have a single service, containing all their functions, for their entire project. As the number of functions grows, it become very difficult to manage configuration. Variables allow you to keep each function config in a separate file, for example sitting next to the handler in the service directory structure. Here is an example of what a serverless.yml might look like in this case:
```

Comment:
Two notes with the first sentence here:

1) Simplify the beginning -- `Use cases exist where users make the decision` to `Some users decide`

2) At the end -- `when working the Serverless.` --> `when working with the Serverless Framework.`